src/classes/AutoNumberService.cls

```diff
@@ -0,0 +1,326 @@
+            an.Starting_Number__c =
+                    Integer.valueOf(fields.get(String.valueOf(
+                            AutoNumber__c.Starting_Number__c)));
+            an.Description__c =
+                    (String) fields.get(String.valueOf(
+                            AutoNumber__c.Description__c));
+            an.Object_API_Name__c =
+                    (String) fields.get(String.valueOf(
+                            AutoNumber__c.Object_API_Name__c));
+            an.Field_API_Name__c =
+                    (String) fields.get(String.valueOf(
+                            AutoNumber__c.Field_API_Name__c));
+            return an;
+        }
+
+        private void validate() {
+            if (!isValidPattern) {
+                throw new AutoNumberException('Invalid Display Format: ' + record
+                        .Display_Format__c);
+            }
+            if (isDuplicate) {
+                throw new AutoNumberException('Duplicate Display Format: ' +
+                        record.Display_Format__c + ' for Object: ' + record
+                        .Object_API_Name__c);
+            }
+        }
+
+        private Boolean isValidPattern {
+            get {
+                return Pattern.matches('.*\\{0*\\}', record.Display_Format__c) &&
+                        record.Display_Format__c.countMatches('{') == 1 &&
+                        record.Display_Format__c.countMatches('}') == 1;
+            }
+        }
+
+        private Boolean isDuplicate {
+            get {
+                for (AutoNumber__c autoNumber : autoNumberService.getAutoNumbers(sObjType)) {
+                    if (autoNumber.Display_Format__c.substringBefore('{') ==
+                            record.Display_Format__c.substringBefore('{')) {
+                        return true;
+                    }
+                }
+                return false;
+            }
+        }
+
+        private Integer getNextAutoNumber() {
+            Integer maxUsedNumber = record.Max_Used_Number__c != null ?
+                    Integer.valueOf(record.Max_Used_Number__c) : 0;
+            Integer startingNumber = record.Starting_Number__c != null ?
+                    Integer.valueOf(record.Starting_Number__c) : null;
+            Integer nextNumber = maxUsedNumber + 1;
+            if (startingNumber != null) {
+                return maxUsedNumber < startingNumber ?
+                        startingNumber :
+                        nextNumber;
+            } else {
+                return nextNumber;
+            }
+        }
+
+        private Integer getNumberLength() {
+            String oB = '{';
+            String cB = '}';
+            Integer a = displayFormat.indexOf(oB);
+            Integer b = displayFormat.indexOf(cB);
+            Integer numLen = b - a - 1;
+            return numLen < 0 ? 0 : numLen;
+        }
+    }
+
+    private AutoNumber getAutoNumber() {
+        List<AutoNumber__c> autoNumbers = [
+                SELECT
```

Comment:
Minor formatting suggestion. I think the first field should be inline with SELECT keyword and subsequent fields are 1 tab/4 spaces and AND keywords are 1 tab/4 spaces.